### ISSUE TITLE
Current master has undefined symbol

### ISSUE THREAD

**Issue by khuck** [ornladios/ADIOS2#2261]
We ask the community to always report bugs. Please open a new issue if some other comes up in the conversation. It will help us stay on topic and manage tasks more efficiently.

**Describe the bug**
Current master of ADIOS2 is failing with an undefined symbol:
```
TAU: Failed loading libTAU-adios2-plugin.so plugin with error: /gpfs/alpine/world-shared/gen010/tau2/summit-adios2-2020.05.15/ibm64linux/lib/shared-pgi-19.9-papi-mpi-pthread-pdt-adios2-pgi/libTAU-adios2-plugin.so: undefined symbol: _ZN6adios22IO12AddTransportESsRKSt3mapISsSsSt4lessISsESaISt4pairIKSsSsEEE
```
which resolves to:
```
adios2::IO::AddTransport(std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::map<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&)
```

Does this symbol need to be exported?  Is it declared without a definition?

**To Reproduce**
A minimal reproducible example is preferred.
Or Steps to reproduce the behavior:
1. configure TAU to use ADIOS2 (see https://github.com/UO-OACISS/tau2/wiki/Using-TAU-to-Profile-and-or-Trace-ADIOS and use the -adios flag or +adios2 with spack - contact @khuck for details if necessary)
2. Run an executable with `tau_exec -adios2 ...`
3. ...

**Expected behavior**
Up until recently, this worked fine.  The only difference was an updated ADIOS2 to use master.

**Desktop (please complete the following information):**
 - OS/Platform: Summit
 - Build PGI 19.9

**Additional context**
for WDM project - measuring XGC+GENE

**Following up**
Was the issue fixed? Please report back.

**Comment by eisenhauer** [comment]
This is a bit of an odd one.  Have you tried to see if you can run an executable *without* TAU?  We're running hundreds of tests on a couple of dozen platforms with every merge, all of which require running executables.  So it seems likely that whatever is going wrong might be external to ADIOS.  Maybe an old library version that is found when running under TAU for some reason?

**Comment by khuck** [comment]
I think it's more likely it's related to PGI than TAU... I am currently doing a git bisect on ADIOS2, and the most recent commit of ADIOS2 that works for me is after #2086 was merged on March 31.  I'll let you know when I narrow it down to the change that seems to have done it...  @suchyta1 had found it first, I was able to reproduce it.  Still bisecting, though....

**Comment by eisenhauer** [comment]
Well, for what it's worth, I tried to look at it from the "did anything change with IO::AddTransport()" point of view, but that interface hasn't changed for years.  However AddTransport() is a simple call and it doesn't seem to match the demangled interface you give, so a compiler issue is maybe indicated...

**Comment by khuck** [comment]
> AddTransport() is a simple call and it doesn't seem to match the demangled interface you give

Not necessarily - AddTransport has two arguments, a string and a Params, and Params is defined as a map from a string to a string so once it gets expanded (map has a key, value, comparator, allocator), it has that bug ugly signature...

**Comment by eisenhauer** [comment]
Hmm.  Can you run executables outside of TAU? Like just "bin/hello_bpWriter" in the ADIOS2 build directory?  Or does that die similarly?

**Comment by khuck** [comment]
that segfaulted, but I assume that's because I tried to run it on a login node of summit.  bpls ran OK (with no arguments, it just exited).

**Comment by suchyta1** [comment]
@eisenhauer @khuck
Like Kevin said with `bpls`, executables do run OK (no segmentation faults) for me if I'm not using `tau_exec`.

**Comment by suchyta1** [comment]
When I do a run with `tau_exec` where the software has been built with GCC-9.1.0, it seems to be hanging at startup (different manifestation of same issue?), so I think something similar persists from PGI to GCC.

**Comment by khuck** [comment]
OK, here's what git bisect says is the bad commit:

```
[khuck@login5.summit ADIOS2]$ git bisect good
1984b530e1c6574fe93b41dc5429eb49e8371c73 is the first bad commit
commit 1984b530e1c6574fe93b41dc5429eb49e8371c73
Author: Brad King <brad.king@kitware.com>
Date:   Wed Mar 4 14:57:56 2020 -0500

    bindings: Split Fortran MPI- and non-MPI bindings into separate libraries
    
    Split the `adios2_f` library into new `adios2_fortran{,_mpi}` libraries.
    
    Preserve the original public-facing `adios2_f` target name as an
    interface library that links both `adios2_fortran` and
    `adios2_fortran_mpi` so that clients using that name still get both.

:040000 040000 7ffaaca400503af936344b6ef1db19e0078ad241 50de11e54c693993d0f1de12f82a809dc009d7ac M	bindings
:040000 040000 c81a06b050d66c8a12c4785efc63d6eb0de32c5b 5d15e2d026eecbddc1dc6fa09753584ea8cf08fa M	cmake
:040000 040000 cf064ae85c0d17cbe379885700527c2ba7d4a2ad 89fcc1f52315930aabd6b58213fb663369de224c M	testing
```

**Comment by eisenhauer** [comment]
Hmm.   I'm going to step back, because I'm not at all sure what might be going on here.  @bradking ?  Possible complexity with Fortran submodules causing weirdness with Tau?

**Comment by khuck** [comment]
To add a bit more context, TAU gets the adios libraries by calling `adios2-config --c-flags` and `adios2-config --c-libs`.  Would this change somehow have an effect such that there are now more/split ADIOS libraries, and `adios2-config --c-libs` output needs to be updated?

**Comment by eisenhauer** [comment]
Tagging @chuckatkins as well...

**Comment by khuck** [comment]
Here's the output I get on Summit when I build with PGI:

```
[khuck@login5.summit ADIOS2]$ ./install-xgc/bin/adios2-config --c-libs
-Wl,-rpath,/autofs/nccs-svm1_home1/khuck/src/ADIOS2/install-xgc/lib64 /autofs/nccs-svm1_home1/khuck/src/ADIOS2/install-xgc/lib64/libadios2_c_mpi.so.2.5.0 /autofs/nccs-svm1_home1/khuck/src/ADIOS2/install-xgc/lib64/libadios2_c.so.2.5.0 -Wl,-rpath-link,/autofs/nccs-svm1_home1/khuck/src/ADIOS2/install-xgc/lib64 
[khuck@login5.summit ADIOS2]$ ./install-xgc/bin/adios2-config --cxx-libs
-Wl,-rpath,/autofs/nccs-svm1_home1/khuck/src/ADIOS2/install-xgc/lib64 /autofs/nccs-svm1_home1/khuck/src/ADIOS2/install-xgc/lib64/libadios2_cxx11_mpi.so.2.5.0 /autofs/nccs-svm1_home1/khuck/src/ADIOS2/install-xgc/lib64/libadios2_cxx11.so.2.5.0 -Wl,-rpath-link,/autofs/nccs-svm1_home1/khuck/src/ADIOS2/install-xgc/lib64 
[khuck@login5.summit ADIOS2]$ ./install-xgc/bin/adios2-config --fortran-libs
-Wl,-rpath,/autofs/nccs-svm1_home1/khuck/src/ADIOS2/install-xgc/lib64 /autofs/nccs-svm1_home1/khuck/src/ADIOS2/install-xgc/lib64/libadios2_f.so 
```

TAU is C++ code, but we use --c-flags because at one point there were some odd flags coming from --cxx-flags that caused us problems.  It looks like I need to switch back to --cxx-flags?

**Comment by khuck** [comment]
I remember now... the `adios2-config --cxx-flags` output includes ` --c++11 --gnu_extensions ` which causes us problems.  So we switched to `--c-flags`, and also `--c-libs` (for consistency).  We only add the c++11 flag for the files that contain ADIOS2 calls (long story, but keep in mind TAU is a 25+ year-old code base...)

I changed to `--c-flags` and `--cxx-libs` and things are working again.  I will do some more tests, then close this if I don't have any more issues.  Thanks!

**Comment by bradking** [comment]
> ```console
> $ ./install-xgc/bin/adios2-config --fortran-libs
> .../ADIOS2/install-xgc/lib64/libadios2_f.so 
> ```

That looks like output from before the bisect result commit `1984b530e1c6574fe93b41dc5429eb49e8371c73`.

Are you sure that is the correct commit?  It shouldn't have affected the `--{c,cxx}-{flags,libs}` output at all.

**Comment by khuck** [comment]
> ```
> ```
> $ ./install-xgc/bin/adios2-config --fortran-libs
> .../ADIOS2/install-xgc/lib64/libadios2_f.so 
> ```
> 
> 
> 
> That looks like output from before the bisect result commit 1984b530e1c6574fe93b41dc5429eb49e8371c73.
> Are you sure that is the correct commit?  It shouldn't have affected the --{c,cxx}-{flags,libs} output at all.
> ```

I might have run those commands against a build of current master.

**Comment by khuck** [comment]
Switching from `adios2-config --c-libs` to `adios2-config --cxx-libs` fixed my issues, so I'm closing this issue.
